File: crates/posthaste-domain/src/model/appearance.rs
use serde::{Deserialize, Serialize};

/// App-level appearance/theme preferences — the renderer's visual settings,
/// stored in `[appearance]` of `app.toml` as the single source of truth (moved
/// out of the opaque `localStorage` "client-preferences" snapshot).
///
/// The backend treats appearance as **pass-through storage**: it does not
/// interpret theme values (the renderer applies them). The enums give the
/// OpenAPI schema self-documentation and reject typos at the parse boundary.
///
/// @spec docs/eph/DESIGN-L2-appearance-toml
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct Appearance {
    pub mode: Option<ThemeMode>,
    pub palette_preset: Option<PalettePresetId>,
    pub density: Option<UiDensity>,
    pub accent_hue: Option<u32>,
    pub glass_theme: Option<GlassTheme>,
}

/// Theme mode: light, dark, or follow the OS preference.
///
/// @spec docs/eph/DESIGN-L2-appearance-toml
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub enum ThemeMode {
    Light,
    Dark,
    System,
}

/// UI density (information density of the layout).
///
/// @spec docs/eph/DESIGN-L2-appearance-toml
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub enum UiDensity {
    Compact,
    Cozy,
    Comfortable,
}

/// Palette preset identifier (the renderer resolves these to concrete palettes).
/// Matches the renderer's `PalettePresetId` set; extend both together when a
/// preset is added.
///
/// @spec docs/eph/DESIGN-L2-appearance-toml
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub enum PalettePresetId {
    Neutral,
    PaperInk,
    Brutalist,
    Glass,
    Acid,
    Marzipan,
    Botanical,
}

/// Advanced glass-theme parameters: a set of decorative "blooms" rendered as the
/// background. Pass-through storage; the renderer normalizes/clamps values.
///
/// @spec docs/eph/DESIGN-L2-appearance-toml
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct GlassTheme {
    pub blooms: Vec<GlassBloom>,
}

/// One decorative bloom in a [`GlassTheme`].
///
/// @spec docs/eph/DESIGN-L2-appearance-toml
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct GlassBloom {
    pub id: String,
    pub hue: u32,
    pub x: f64,
    pub y: f64,
    pub opacity: f64,
    pub radius: f64,
}

File: apps/web/src/api/types/appearance.ts
/**
 * App-level appearance/theme preferences — the wire shape of `[appearance]` in
 * `app.toml`. TOML is the source of truth; the renderer keeps a derived
 * `localStorage` mirror only for fast boot (no FOUC). These types model the
 * daemon wire schema exactly.
 *
 * @spec docs/eph/DESIGN-L2-appearance-toml
 */
export type ThemeMode = 'light' | 'dark' | 'system'
export type UiDensity = 'compact' | 'cozy' | 'comfortable'
export type PalettePresetId =
  | 'neutral'
  | 'paperInk'
  | 'brutalist'
  | 'glass'
  | 'acid'
  | 'marzipan'
  | 'botanical'

/** One decorative bloom in a {@link GlassTheme}. */
export interface GlassBloom {
  id: string
  hue: number
  x: number
  y: number
  opacity: number
  radius: number
}

/** Advanced glass-theme parameters: decorative "blooms" rendered as the background. */
export interface GlassTheme {
  blooms: GlassBloom[]
}

/** UI appearance prefs. All fields optional; an absent/cleared field means "use the renderer default". */
export interface Appearance {
  mode?: ThemeMode | null
  palettePreset?: PalettePresetId | null
  density?: UiDensity | null
  accentHue?: number | null
  glassTheme?: GlassTheme | null
}
